TST: Add testing for hyp2f1 for complex values in anticipation of move to Cython.

steppi opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
This closes no issues but is a preliminary to future work that will address gh-7340, gh-8054, gh-8083, gh-4894, as well as other defects in hyp2f1 for complex numbers for which there are no existing issues.
#### What does this implement/fix?
<!--Please explain your changes.-->
This PR is a follow up to gh-14256, which was an attempt to begin translating hyp2f1 for complex values into Cython to improve maintainability and make it easier to fix the many longstanding issues with this function. It was decided that the previous PR was too unwieldy for prompt review so this is the first in a series that break the previous PR into several smaller ones. As explained below, although the line count of this PR also appears unwieldy, a large bulk of the lines are dedicated to test cases and others to explanatory comments. I've taken steps to save the reviewer from the need to validate each test case individually. These steps are described below. This PR does not contain specialized numerical mathematics and should be accessible to a wider class of reviewers than the previous PR.

Tests are added for hyp2f1 for complex values for a large but not exhaustive set of the special cases that need to be handled when computing this function. Additional tests covering additional cases will be added incrementally as progress is made translating the existing Fortran implementation of hyp2f1 into Cython. The script `scipy.special._precompute.hyp2f1_data.py` has also been added which compares scipy's hyp2f1 against mpmath's in a grid of points in the complex plane over a range of parameter combinations. This script is useful for assessing the health of hyp2f1, ensuring that no regressions occur when translating into Cython, determining where fixes need to be made, and deciding between different options when making changes to the implementation.


#### Additional information
<!--Any additional information you think is important.-->
Test cases are written in a verbose style using `namedtuple`s and `pytest.mark.parameterize`. 

For example

```python
    @pytest.mark.parametrize(
        "hyp2f1_test_case",
        [
            pytest.param(
                Hyp2f1TestCase(
                    a=0.5,
                    b=0.2,
                    c=-10,
                    z=0.2 + 0.2j,
                    expected=np.inf + 0j,
                    rtol=0
                )
            ),
            pytest.param(
                Hyp2f1TestCase(
                    a=0.5,
                    b=0.2,
                    c=-10,
                    z=0 + 0j,
                    expected=1 + 0j,
                    rtol=0
                ),
                marks=pytest.mark.xfail(reason="gh-7340")
            ),
...
```


I think this helps with readability. It also makes it easy to mark individual test cases as expected to fail. These failing tests cases demonstrate flaws in the current implementation that will be addressed in future PRs. A large portion of the lines in this PR come from these verbosely expressed test cases. A function `scipy.special._precompute.hyp2f1_data.make_hyp2f1_test_cases` has been added which can help generate the boilerplate for these test cases.

The following steps have been taken to spare the reviewer the tedium of having to check the validity of all test cases.

1. Assertions have been added that ensure that a test case's parameters actually satisfy the conditions for the corresponding test. For example
```python
    def test_region2(self, hyp2f1_test_case):
        """|z| < 1 and real(z) < 0."""
        a, b, c, z, expected, rtol = hyp2f1_test_case
        assert abs(z) < 1 and z.real < 0  # Tests the test
        assert_allclose(hyp2f1(a, b, c, z), expected, rtol=rtol)
```
2. A test, `test_test_hyp2f1`, has been added which checks that the expected value for each test case is actually the value that is computed by mpmath. This is written in such a way as to ensure there's a single source of truth for the test cases. It has been marked as slow so that it will be skipped on the continuous integration builds (as far as I'm aware). It will also be skipped if mpmath is not available. At this time, this test is not actually slow because there aren't that many test cases. This test may become slow in the future if more test cases are added.

rgommers commented on scipy/special/tests/test_hyp2f1.py:
```diff
@@ -0,0 +1,608 @@
+    # Marked as slow so it won't run on continuous integration builds.
+    # This test is not actually that slow, though could become slow in
+    # the future if enough test cases are added.
+    @pytest.mark.slow
```
Note that this will run on CI jobs that have `full` in the name - and the Windows full test suite run is already at around the 60 min timeout limit. There's a `pytest.mark.xslow` you want to use instead.

steppi replied:
Thanks, I didn't realize that. I'll change it.  The entire test suite currently runs in 0.39 seconds on my machine, but this last test could creep up in running time as more test cases are added. I added `test_test_hyp2f1.py` at the end because I was worried reviewers could get bogged down trying to check the validity of each test case individually. I wanted there to be a convenient way for reviewers to check all of the cases. The problem then reduces to checking that the logic to gather the test cases out of the pytest marks is correct.

steppi replied:
I've looked into it, and I think this test actually should be marked slow, not xslow. Code authors and reviewers would not want it to be skipped,and it's reasonable to run in full builds. It should be skipped by default. It's status is similar to the tests marked slow in `test_mpmath.py`. The running time should never exceed the 1-10 second range.

rgommers replied:
Okay, thanks for looking into that!

rgommers commented on scipy/special/tests/test_hyp2f1.py:
```diff
@@ -0,0 +1,608 @@
+    match the situations that are intended. A final test `test_test_hyp2f1`
+    checks that the expected values in the test cases actually match what
+    is computed by mpmath. This test is marked slow even though it isn't
+    particularly slow so that it won't run on continuous integration builds.
```
Shouldn't the tests in this class be skipped if the data file doesn't exist? If that is done somewhere, I'm failing to spot it.

steppi replied:
Ah, I should have been more clear here. There is actually no data file used. This test is gathering the test cases out of the pytest marks and recomputing the expected value with mpmath to check if it matches. I was little worried that this just a little too clever in the bad sense though. I'm also not sure if the structure of the pytest marks could somehow change down the line, breaking this test, making me feel it's a little brittle.

steppi replied:
Oh, I understand the cause of the confusion now. The test imports something called `mp_hyp2f1` from `scipy.special._precompute.hyp2f1_data`. It's more reasonable than not to expect this to do something with precomputed values, due to the name of the module, and because one should expect the test just use to `mpmath.hyp2f1` directly. The only source of precomputed values I've given is an enormous file generated by that script.

What's really happening is that `mp_hyp2f1` is a modification of `mpmath.hyp2f1` that returns the same value as `scipy.special.hyp2f1` on the branch cut, where choice of value is discretionary. I placed `mp_hyp2f1` in the file for the precompute script because at first that was the only place it was used. When I later added `test_test_hyp2f1`, I imported it from _precompute, which is pretty confusing now that I think about it.

Is there a better place to put this function? For now I've just added a comment to `test_hyp2f1.py` explaining what the function does.

rgommers replied:
> I'm also not sure if the structure of the pytest marks could somehow change down the line, breaking this test, making me feel it's a little brittle. Edit: I did a quick study of the pytest documentation, and the source code and its git history. It's given me some confidence that this way of getting the parameters out of the test cases should be pretty stable.

I think that's fine then. We have a lot of tests that use this kind of parameter parsing (mostly custom rather than `pytest`-based, because it predates our migration to pytest). And even if it breaks at some point, it'll likely do so loudly so we can just patch it up.

rgommers replied:
> Is there a better place to put this function? For now I've just added a comment to `test_hyp2f1.py` explaining what the function does.

If the comment is clear enough it's fine. Otherwise, you could have turned it around? Put `mp_hyp2f1` into the `test_` file and make the `_precompute` functionality import it from there?

steppi replied:
I moved it around to put `mp_hyp2f1` in the test and imported it in `_precompute` from there. I was cargo culting some nonsense about not importing from the test directory, but it seems fine to do it here and much better than the alternative.

rgommers replied:
This looks much better, thanks!

rgommers commented:
Thanks @steppi, this looks really good and thorough. My main question right now is about the test strategy. We have `slow` and `xslow` decorators, the former is run on a subset of CI jobs (those with `full` in the name, can also be run with `scipy.test('full')`) and the latter need to be enabled with `export SCIPY_XSLOW=1`.

These `hyp2f1` tests are really specific though. We don't want regular contributors to have to generate this 700 MB file, nor do we want to commit it to the repo. Given how heavy it is, it's a development tool rather than a test suite addition I'd say. So it may be better to put it behind a separate switch (say a `SCIPY_MPMATH_XSLOW` env var) than adding it to `xslow`.

It seems a little tricky to get the right balance here for the test suite. I'd say the division is:
- regular tests: fast individual tests (sub-second), total tests for the feature on the order of seconds. We want to run O(40,000) tests in a couple of minutes here. Coverage should be enough to not badly break anything.
- full test suite (includes `slow`): these aren't run by default with `scipy.test()`, but they are run in CI jobs for each architecture. So the total runtime should still be limited to ~20-30 minutes on CI. Individual tests can take in the 1-10 second range here.
- tests marked `xslow`. Not run in CI, and enables/disables some benchmarks too IIRC. Used for special cases - tests taking in the 10 sec - 30 min range, high-memory tests, etc. 

For something that takes 40 minutes on an 8-core machine, I'd say the only comparable thing we have is the `optimize` global benchmarks suite (which is controlled by `SCIPY_XSLOW` currently).

steppi commented:
@rgommers

Sorry, I should have been more explicit about what's going on. The script,  `scipy.special._precompute.hyp2f1_data.py` shouldn't be run unless a user explicitly runs it. The data file is not meant to be checked into version control. What I did was put a command line option to specify where the output file will go, so that a user can put it somewhere on their local machine if they want. My workflow was to run this first on the original Fortran implementation to gather a baseline and then again on my translated code. I would then explore the data in the file with pandas to find cases where either the translation was less precise than the original or vice versa and add these to test cases. Then I would work on improving the translation so that all of the cases would pass. There's a function in `scipy.special._precompute.hyp2f1_data.py` that generates the pytest boilerplate for the test cases, which I then paste into `scipy.special.tests.test_hyp2f1.py`.

rgommers commented:
This looks very clean now. Test structure wise I'm happy. I haven't tried to wrap my head around the functionality under test yet, since @person142 understands all that way better. But if it is necessary I'll do that (maybe next weekend), this kind of PR we can get merged in a reasonable amount of time.

steppi commented:
Great, thanks. I plan to make the future PRs for hyp2f1 as small and atomic as is reasonable, changing things bit by bit until none of the old hyp2f1 remains. Optimistically, I'd hope to be able to get through everything in around 1.5 years. I read your post on [The cost of open source software](https://rgommers.github.io/2019/06/the-cost-of-an-open-source-contribution/) and will try to keep what you said there in mind and not create unnecessary burden for the maintainers. I can probably help out with some amount of review work along the way.

person142 commented:
I should be able to get to this in the second half of this week-no time until after Tuesday for sure though.

rgommers commented:
> Optimistically, I'd hope to be able to get through everything in around 1.5 years.

Ouch, that's a long road ahead.

> I read your post on [The cost of open source software](https://rgommers.github.io/2019/06/the-cost-of-an-open-source-contribution/) and will try to keep what you said there in mind and not create unnecessary burden for the maintainers. I can probably help out with some amount of review work along the way.

That's always nice to hear:) I'm not worried about it here though - your PRs and explanations look great, I wish they were all like this!

steppi commented:
> > Optimistically, I'd hope to be able to get through everything in around 1.5 years.
> 
> Ouch, that's a long road ahead.
> 
Ah. It's really not too bad compared with the length of time some of the outstanding PRs for hyp2f1 have been open. I have a roadmap planned out and will work on it a little at a time and try to keep a steady trickle of small PRs coming.

> > I read your post on [The cost of open source software](https://rgommers.github.io/2019/06/the-cost-of-an-open-source-contribution/) and will try to keep what you said there in mind and not create unnecessary burden for the maintainers. I can probably help out with some amount of review work along the way.
> 
> That's always nice to hear:) I'm not worried about it here though - your PRs and explanations look great, I wish they were all like this!

Thanks! That's high praise :)

steppi commented:
I've added optional arguments to adjust the number of grid points used, restrict the regions for the arguments `z`, and restrict the parameter groups. The regions and parameter groups are described in the module level Docstring for `scipy.special._precompute.hyp2f1_data.py`. I added these while working on the fix for gh-8054 while wanting to compute with a larger grid size, but only keeping the relevant regions and parameter groups. By the way, I'm mostly done with the fix for gh-8054. The series from [this paper](https://arxiv.org/abs/1306.2046) seems to work well, but there are still small adjustments to make. I also have two more PRs queued up that incorporate the rest of the changes from my closed PR, gh-14256.

steppi commented:
@person142 Do you think you'll have time to look at this in the next couple weeks?

person142 commented on scipy/special/tests/test_hyp2f1.py:
```diff
@@ -0,0 +1,649 @@
+    return complex(mpmath.hyp2f1(a, b, c, z_mpmath))
+
+
+Hyp2f1TestCase = namedtuple(
```
Recommend the new named tuple syntax; namely:

```python
from typing import NamedTuple

class Hyp2f1TestCase(NamedTuple):
    a: float
    b: float
    ...
```

(Or use dataclasses.)

steppi replied:
Thanks for the suggestion. Now that I've seen it, I like the new syntax better too.

person142 commented on scipy/special/_precompute/hyp2f1_data.py:
```diff
@@ -0,0 +1,395 @@
+        default=None,
+        help="Restrict to supplied parameter groups. See the Docstring for"
+        " this module for more info on parameter groups. Calculate for all"
+        " paremeter groups by default."
```
typo "parameter"

steppi replied:
Oh, oops.

person142 reviewed (approved):
Ahhhhh sorry it took me so long to review.

Looks really good; love how thorough it is. Left some very small comments; will give some time if you want to address o/w ping me and I'll merge.

steppi commented:
Thanks @person142. I've made the suggested changes. I have another one ready to go that I'll submit tomorrow. No worries on the time. If we can get through one small PR per month, I think we can get all of hyp2f1's problems resolved in about a year without either of us eating too much into our spare time.
